modules/concurrency/include/hephaestus/concurrency/spinner.h

```diff
@@ -13,22 +13,37 @@ namespace heph::concurrency {
 
 /// A spinner is a class that spins in a loop calling a user-defined function.
 /// If the function is blocking, the spinner will block the thread.
-/// If the input `rate_hz` is set to a non-zero value, the spinner will call the user-defined function at
-/// the given fixed rate.
+/// If the input `rate_hz` is set to a non-infinite value, the spinner will call the user-defined function at
+/// the given fixed rate. The spinner behavior can be configured using callbacks.
 class Spinner {
 public:
   enum class SpinResult : bool { CONTINUE, STOP };
   using StoppableCallback = std::function<SpinResult()>;
   using Callback = std::function<void()>;
 
-  /// @brief Create a spinner with a stoppable callback. A stoppable callback is a function that returns
-  /// SpinResult::STOP to indicate that the spinner should stop.
-  /// Example: a callback that stops after 10 iterations.
-  explicit Spinner(StoppableCallback&& stoppable_callback, double rate_hz = 0);
+  /// @brief Create a callback for the spinner which never stops.
+  [[nodiscard]] static auto createNeverStoppingCallback(Callback&& callback) -> StoppableCallback;
+
+  struct StateMachineCallbacks {
+    using TransitionCallback = std::function<void()>;
+    using PolicyCallback = std::function<bool()>;
+
+    TransitionCallback init_cb = []() {};       //!< Handles initialization.
+    TransitionCallback spin_once_cb = []() {};  //!< Handles execution.
+
+    PolicyCallback shall_stop_spinning_cb = []() { return false; };  //!< Default: spin indefinitely.
+    PolicyCallback shall_restart_cb = []() { return false; };        //!< Default: do not restart.
+  };
 
-  /// @brief Create a continuous spinner which cannot be stopped by the callback.
-  /// Example: a callback which reads data from a sensor, until the spinner is stopped.
-  explicit Spinner(Callback&& callback, double rate_hz = 0);
+  /// @brief Create a callback for the spinner which internally handles the state machine.
+  [[nodiscard]] static auto createCallbackWithStateMachine(StateMachineCallbacks&& callbacks)
+      -> StoppableCallback;
+
+  /// @brief Create a spinner with a stoppable callback. A stoppable callback is a function that returns
+  /// SpinResult::STOP to indicate that the spinner should stop. Other types of callbacks are supported via
+  /// mappings to StoppableCallback. Example: a callback that stops after 10 iterations.
+  explicit Spinner(StoppableCallback&& stoppable_callback,
```

Comment:
Question: didn't we decide that we actually do not need the stoppable callback as we can use `shall_stop_spinning_cb` for this case? In case we use the stoppable callback, shouldn't we remove `shall_stop_spinning_cb`?

Reply:
The stoppable callback is needed to separate the state machine from the spinner.

The spinner spins forever (`Spinner::SpinResult::Continue`), until it is told by the callback to exit (`Spinner::SpinResult::Stop`). We don't impose a reason for the stop, which can either be a simple StoppableCallback where the iteration counter has elapsed, or it can be a callback stemming from a state machine, where the state machine reached exit.